### PR TITLE
Confirming case distinction

### PR DESCRIPTION
**Pull request by psteyer**
### Description

Addressing Issue #364 

- updating regex to handle more edge cases 
- adding test

### CODE REVIEW

**Review by schuylermartin45 — CHANGES_REQUESTED** [review]

File: tests/commands/test_bump_recipe.py [hunk 1]
```diff
@@ -513,3 +513,28 @@ def test_new_line_removal(
...
+def test_pypi_url_regex_matching(url: str, should_match: bool) -> None:
+    """
+    Test that the PYPI_URL regex pattern correctly identifies PyPI source URLs.
+
+    :param url: URL to test against the PyPI regex pattern
+    :param should_match: Whether the URL should match the PyPI pattern
+    """
+    # pylint: disable=protected-access
+    pypi_match = bump_recipe._Regex.PYPI_URL.match(url)
+    assert (pypi_match is not None) == should_match
```
Comment (schuylermartin45) [inline tests/commands/test_bump_recipe.py:540]: Generally speaking, we shouldn't be testing private members. Private members are expected to change. It is the public API that we want to ensure that works.

In the future, for all we know, we may implement this URL check some other way. Then we will cause more work for ourselves because we used a bunch of non-public features.

I think we should write this in the style of the other tests that uses mocked URLs and mocked test files. Arguably it isn't really a unit test at that point, and more just a test of the CLI, but it should be more stable and follows more "best practices".

Comment (schuylermartin45) [reply tests/commands/test_bump_recipe.py:550]: I believe this comment still applies. I think we should get rid of this test in favor of the other tests.

File: conda_recipe_manager/commands/bump_recipe.py [hunk 2]
```diff
@@ -379,36 +378,17 @@ def _correct_pypi_url(recipe_reader: RecipeReader, url_path: str) -> tuple[str,
     #     changes, less disruption/confusion for our package builders (or so the theory goes).
     #  2) The PyPi API generally produces a "post-redirect" URL for package artifacts. Again, as to not change the file
     #     dramatically, we want to keep the commonly used `https://pypi.io` URL.
```
Comment (schuylermartin45) [inline conda_recipe_manager/commands/bump_recipe.py:380]: I can't select the whole comment in the diff, but this comment needs to be updated with the reasoning behind the new logic. We should also link-back to the original ticket for context.

File: conda_recipe_manager/commands/bump_recipe.py [hunk 3]
```diff
@@ -379,36 +378,17 @@ def _correct_pypi_url(recipe_reader: RecipeReader, url_path: str) -> tuple[str,
...
-            ),
-        )
-        return (fetcher_url, f"{base_original_url}/{filename_with_var}")
+    filename: Final = pypi_meta.releases[version_value].filename
+    name: Final = pypi_meta.info.name
+    new_url: Final = f"https://pypi.org/packages/source/{ name[0] }/{ name }/{ filename }"
 
-    return (fetcher_url, f"{base_original_url}/{filename}")
+    if new_url is None:
+        raise FetchError("Unable to determine download url for PyPi API request.")
```
Comment (schuylermartin45) [inline conda_recipe_manager/commands/bump_recipe.py:386]: This check needs to happen sooner has to check if `name` and `filename` are both _non-empty_. `new_url` uses both of these variables and will produce a pretty useless URL if either or both variable is empty.

Personally I would do something like:
```
if not (filename and name):
  raise FetchError

new_url = ...
return return_url
```

Also accessing `""[0]` will cause an index exception.

**Review by schuylermartin45 — COMMENTED** [review]
A few small notes. Also did we ever try running the `crm bump-recipe` to recreate the original issue locally? It might not be a bad idea to ensure it works in the real world at least once against that `Cerberus` package.

File: conda_recipe_manager/commands/bump_recipe.py (hunk 2 above)
Comment (psteyer) [reply conda_recipe_manager/commands/bump_recipe.py]: Added a comment, Hopefully it makes enough sense. Added issue ticket link also.

File: conda_recipe_manager/commands/bump_recipe.py [hunk 4]
```diff
@@ -374,41 +373,21 @@ def _correct_pypi_url(recipe_reader: RecipeReader, url_path: str) -> tuple[str,
...
-            version_value,
-            (
-                # NOTE: The double-escaping of the outer `{{` and `}}` braces.
-                f"{{{{ {_RecipeVars.VERSION} }}}}"
-                if recipe_reader.get_schema_version() == SchemaVersion.V0
-                else f"${{{{ {_RecipeVars.VERSION} }}}}"
-            ),
-        )
-        return (fetcher_url, f"{base_original_url}/{filename_with_var}")
+    # Using the PyPI metadata we can construct a url that will work for the fetcher. Incase the recipe name is
```
Comment (schuylermartin45) [inline conda_recipe_manager/commands/bump_recipe.py:376]:
```suggestion
    # Using the PyPI metadata we can construct a url that will work for the fetcher. In case the recipe name is
```
Typo

**Review by schuylermartin45 — APPROVED** [review]
I committed a few little comment changes, but I think this looks good now. I'll merge when this passes.
